### PR TITLE
[WIP] Add MTP dummy_run and Adapt torchair graph mode

### PR DESCRIPTION
**Pull request by JC-ut0**
### What this PR does / why we need it?

1. Add MTP dummy_run
2. Adapt torchair graph mode， this draft still have accuracy bug with torchair [WIP]

### CODE REVIEW

**Comment by mengwei805** [comment]
pls rebase all u commit to 1

File: vllm_ascend/models/deepseek_v2.py [hunk 1]
```diff
@@ -478,7 +478,8 @@ def forward(
                 hidden_states_or_q_c = self.q_a_layernorm(ckq)
         else:
             hidden_states_or_q_c = hidden_states
-        if self.torchair_graph_enabled:
+        is_mtp_model = attn_metadata is not None and attn_metadata.is_mtp_model
```
Comment (mengwei805) [inline vllm_ascend/models/deepseek_v2.py:485]: I think attn_metadata adding is_mtp_model attribute is not a good design

File: vllm_ascend/worker/model_runner_v1.py [hunk 2]
```diff
@@ -215,8 +215,13 @@ def __init__(self, vllm_config: VllmConfig, device: torch.device):
...
+        self.actual_seq_q_lens = []
+        self.spec_token_num = 0
+        self.decode_token_per_req = 1
         if self.speculative_config:
             self.use_spec_decode = True
+            self.spec_token_num = self.speculative_config.num_speculative_tokens
+            assert self.spec_token_num > 0
             self.spec_attn_mask = torch.triu(torch.ones(2048,
                                                         2048,
                                                         dtype=torch.bool),
```
Comment (mengwei805) [inline vllm_ascend/worker/model_runner_v1.py:224]: Why is it 2048? This is a devil number.

File: vllm_ascend/attention/mla_v1.py [hunk 3]
```diff
@@ -313,48 +315,62 @@ def _get_graph_runner_block_tables(
...
-        slot_mapping = torch.full((num_reqs, ),
+        slot_mapping = torch.full((num_tokens, ),
                                   PAD_SLOT_ID,
                                   dtype=torch.int32,
                                   device=device)
         query_start_loc = torch.full((num_reqs, ),
                                      -1,
                                      dtype=torch.int32,
                                      device=device)
+        if self.runner.speculative_config.method == 'deepseek_mtp' and not is_mtp_model:
```
Comment (1092626063) [inline vllm_ascend/attention/mla_v1.py]: `self.runner.speculative_config` needs to check if it is `None`, it will get error in Non-mtp scene

File: vllm_ascend/models/deepseek_v2.py (hunk 1 above)
Comment (JC-ut0) [reply vllm_ascend/models/deepseek_v2.py:485]: This is a temporary workaround and will be removed later.

File: vllm_ascend/worker/model_runner_v1.py (hunk 2 above)
Comment (JC-ut0) [reply vllm_ascend/worker/model_runner_v1.py:224]: Accordding to the document: https://www.hiascend.com/document/detail/zh/Pytorch/700/apiref/apilist/ptaoplist_001232.html
when sparse mode is 3, attention mask must be a shape [2048,2048] triu tensor

File: vllm_ascend/attention/mla_v1.py (hunk 3 above)
Comment (JC-ut0) [reply vllm_ascend/attention/mla_v1.py]: solved

File: vllm_ascend/attention/mla_v1.py [hunk 4]
```diff
@@ -971,6 +980,26 @@ def _forward_decode(
                                  self.qk_rope_head_dim)
                 input_layout = "BNSD"
 
+            # TorchAir's shape is [bs, num_heads_per_rank, q_seq_len, dim]
```
Comment (ganyi1996ppo) [inline vllm_ascend/attention/mla_v1.py:983]: Is this comments correct? I see you don't have any transpose operation when you convert the format to TND

File: vllm_ascend/attention/mla_v1.py [hunk 5]
```diff
@@ -971,6 +980,26 @@ def _forward_decode(
...
+            if attn_metadata.attn_state == AscendAttentionState.SpecDecoding:
+                assert num_tokens % self.spec_token_num == 0
+                # [bs * q_seq_len, num_heads_per_rank, dim]
+                input_layout = "TND"
+                q_nope = q_nope.view(num_tokens, self.num_heads, -1)
```
Comment (ganyi1996ppo) [inline vllm_ascend/attention/mla_v1.py:988]: So, you are doing mtp  attention in TND format, therefore, no more conversion here is needed other than some metadata change right?

File: vllm_ascend/worker/model_runner_v1.py [hunk 6]
```diff
@@ -1541,7 +1576,12 @@ def _dummy_run(
         # has num_tokens in total.
         assert num_tokens <= self.scheduler_config.max_num_batched_tokens
         max_num_reqs = self.scheduler_config.max_num_seqs
-        num_reqs = min(num_tokens, max_num_reqs)
+        num_reqs = math.ceil(num_tokens / self.decode_token_per_req)
+        if with_prefill:
+            num_reqs = min(num_tokens, max_num_reqs)
```
Comment (ganyi1996ppo) [inline vllm_ascend/worker/model_runner_v1.py:1581]: Why do this in prefill? looks strange

File: vllm_ascend/attention/mla_v1.py (hunk 5 above)
Comment (JC-ut0) [reply vllm_ascend/attention/mla_v1.py:988]: correct

**Review by ganyi1996ppo — APPROVED** [review]
